src/components/controls/controls.js

```diff
@@ -0,0 +1,225 @@
+      pitch.rotation.x = Math.max(-PI_2, Math.min(PI_2, pitch.rotation.x));
+      el.setAttribute('rotation', {
+        x: THREE.Math.radToDeg(pitch.rotation.x),
+        y: THREE.Math.radToDeg(yaw.rotation.y),
+        z: 0
+      });
+    } else if (control) {
+      el.setAttribute('rotation', control.getRotation());
+    }
+  },
+
+  /**
+   * Updates velocity based on input from the active position component, if any.
+   * @param {number} dt
+   */
+  updateVelocity: function (dt) {
+    var control;
+    var el = this.el;
+    var data = this.data;
+    var velocity = this.velocity;
+
+    if (el.hasAttribute('velocity')) {
+      velocity.copy(el.getComputedAttribute('velocity'));
+    }
+
+    velocity.x -= velocity.x * data.positionEasing * dt / 1000;
+    velocity.z -= velocity.z * data.positionEasing * dt / 1000;
+
+    control = this.getActivePositionControls();
+    if (control && control.getVelocityDelta) {
+      this.applyVelocityDelta(dt, control.getVelocityDelta(dt));
+    } else if (control) {
+      velocity.copy(control.getVelocity());
+    }
+
+    el.setAttribute('velocity', {x: velocity.x, y: velocity.y, z: velocity.z});
+  },
+
+  /**
+   * Updates internal velocity. Takes a velocity delta (relative to current heading), rotates it
+   * to the heading, and calculates velocity in world space.
+   * @param {number} dt
+   * @param {THREE.Vector3} dVelocity
+   */
+  applyVelocityDelta: function (dt, dVelocity) {
+    var data = this.data;
+    var rotation = this.el.getAttribute('rotation');
+
+    // Set acceleration
+    if (dVelocity.length() > 1) {
+      dVelocity.setLength(this.data.positionAcceleration * dt / 1000);
+    } else {
+      dVelocity.multiplyScalar(this.data.positionAcceleration * dt / 1000);
+    }
+
+    // Rotate to heading
+    if (rotation) {
+      this.heading.set(
+        data.flyingEnabled ? THREE.Math.degToRad(rotation.x) : 0,
+        THREE.Math.degToRad(rotation.y),
+        0
+      );
+      dVelocity.applyEuler(this.heading);
+    }
+
+    this.velocity.add(dVelocity);
+  },
+
+  /**
+   * Validates position and rotation controls, asserting that each was registered with
+   * AFRAME.registerControls.
+   * @return {[type]} [description]
+   */
+  validateControls: function () {
+    var i;
+    var name;
+    var el = this.el;
+    var data = this.data;
+    var system = this.system;
+    var missingControls = [];
+
+    for (i = 0; i < data.position.length; i++) {
+      name = data.position[i];
+      if (!system.positionControls[name]) {
+        throw new Error('Component `' + name + '` must be ' +
+                        'registered with AFRAME.registerControls().');
+      } else if (!el.components[name]) {
+        missingControls.push(name);
+      }
+    }
+
+    for (i = 0; i < data.rotation.length; i++) {
+      name = data.rotation[i];
+      if (!system.rotationControls[name]) {
+        throw new Error('Component `' + name + '` must be ' +
```

Comment:
So that's just a bug in my docs, it sounds like – should be `<a-entity controls="rotation: custom-controls1, custom-controls2"/>` with a comma, not just a space. Comma-delimited is working fine for me; fixed the docs.